homeassistant/components/light/greenwave.py

```diff
@@ -0,0 +1,148 @@
+    """Check Device Status."""
+    return 'offline' not in device
+
+
+def gateway_action(host, did, action, value):
+    """Do Action for TCP Connected Gateway."""
+    import requests
+    if action == 'xml':
+        url = ('http://' + host + '/gwr/gop.php?cmd=GWRBatch&data=<gwrcmds>'
+               '<gwrcmd><gcmd>RoomGetCarousel</gcmd><gdata><gip><version>1'
+               '</version><token>1234567890</token><fields>name,status'
+               '</fields></gip></gdata></gwrcmd></gwrcmds>&fmt=xml')
+        resp = requests.get(url)
+        return resp
+    if action == 'level':
+        action = '<type>level</type>'
+    else:
+        action = ''
+    url = ('http://' + host + '/gwr/gop.php?cmd=DeviceSendCommand&data=<gip>'
+           '<version>1</version><token>1234567890</token>'
+           '<did>' + did + '</did>'
+           '<value>' + str(value) + '</value>' + action + '</gip>&fmt=xml')
+    requests.get(url)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup TCP Connected Platform."""
+    host = config.get(CONF_HOST)
+    doc = pull_xml(host)
+    add_devices(GreenwaveLights(device, host) for device in
+                doc['gwrcmds']['gwrcmd']['gdata']['gip']['room']['device'])
+
+
+class GreenwaveLights(Light):
+    """Representation of an TCP Connected Light."""
+
+    def __init__(self, light, host):
+        """Initialize a TCP Connected Light."""
+        self._did = light['did']
+        self._name = light['name']
+        self._state = int(light['state'])
+        self._brightness = get_brightness(light)
+        self._host = host
+        self._online = is_online(light)
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORTED_FEATURES
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._online
+
+    @property
+    def did(self):
+        """Return the device id of this light."""
+        return self._did
+
+    @property
+    def name(self):
+        """Return the display name of this light."""
+        return self._name
+
+    @property
+    def brightness(self):
+        """Return the brightness of the light."""
+        return self._brightness
+
+    @property
+    def is_on(self):
+        """Return true if light is on."""
+        return self._state
+
+    def turn_on(self, **kwargs):
+        """Instruct the light to turn on."""
+        temp_brightness = str(number_range(kwargs.get(ATTR_BRIGHTNESS, 255),
+                                           1, 255, 1, 100))
+        gateway_action(self._host, self._did, 'level', temp_brightness)
+        gateway_action(self._host, self._did, 'on', 1)
+
+    def turn_off(self, **kwargs):
+        """Instruct the light to turn off."""
+        gateway_action(self._host, self._did, 'off', 0)
+
+    def update(self):
+        """Fetch new state data for this light."""
+        doc = pull_xml(self._host)

```

Comment:
We should not fetch ALL light states for each bulb. Instead, create a shared state object. Have a look at the Hue platform.

Reply:
(We should not care what the Android app does, it is wasteful of CPU/bandwidth from the Home Assistant perspective)